pkg/scheduler/plugins/deviceshare/device_cache.go

```diff
@@ -0,0 +1,335 @@
+/*
+Copyright 2022 The Koordinator Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package deviceshare
+
+import (
+	"fmt"
+	"sync"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/types"
+	quotav1 "k8s.io/apiserver/pkg/quota/v1"
+	"k8s.io/klog/v2"
+
+	apiext "github.com/koordinator-sh/koordinator/apis/extension"
+	schedulingv1alpha1 "github.com/koordinator-sh/koordinator/apis/scheduling/v1alpha1"
+)
+
+type nodeDeviceCache struct {
+	lock sync.Mutex
+	// nodeDeviceInfos stores nodeDevice for each node
+	// and uses node name as map key.
+	nodeDeviceInfos map[string]*nodeDevice
+}
+
+type nodeDevice struct {
+	lock        sync.Mutex
+	deviceTotal map[schedulingv1alpha1.DeviceType]deviceResources
+	deviceFree  map[schedulingv1alpha1.DeviceType]deviceResources
+	deviceUsed  map[schedulingv1alpha1.DeviceType]deviceResources
+	allocateSet map[schedulingv1alpha1.DeviceType]map[types.NamespacedName]*corev1.Pod
```

Comment:
Do we need to save the pointer of the Pod? I don't see where the Pod information needs to be used.

Reply:
modified to `allocateSet map[schedulingv1alpha1.DeviceType]map[types.NamespacedName]struct{}`